src/circuit_builder.rs

```diff
@@ -38,44 +273,162 @@ impl PublicInput {
 
 /// Circuit representation for a gadget with all of the tools that it
 /// should implement.
-pub trait Circuit<'a>
-where
-    Self: Sized + Default,
-{
+pub trait Circuit: Sized {
     /// Gadget implementation used to fill the composer.
-    fn gadget(&mut self, composer: &mut StandardComposer) -> Result<Vec<PublicInput>>;
-    /// Compiles the circuit by using a function that returns a `Result`
-    /// with the `ProverKey`, `VerifierKey` and the circuit size.
-    fn compile(&mut self, pub_params: &PublicParameters)
-        -> Result<(ProverKey, VerifierKey, usize)>;
+    fn gadget(&mut self, composer: &mut StandardComposer) -> Result<Vec<PublicInput>, Error>;
+
+    /// Compiles the circuit.
+    ///
+    /// Set the commit, opening, prover and verifier keys.
+    ///
+    /// Also, set the circuit size.
+    fn compile(&mut self, pub_params: &PublicParameters) -> Result<(), Error> {
+        let (ck, ok) = pub_params.trim(Self::max_degree())?;
+
+        self.set_commit_key(ck.clone());
+        self.set_opening_key(ok);
+
+        let mut prover = Prover::new(Self::compile_label());
+        let mut verifier = Verifier::new(Self::compile_label());
+
+        self.gadget(prover.mut_cs())?;
+        self.gadget(verifier.mut_cs())?;
+
+        let size = prover.mut_cs().circuit_size();
+        self.set_circuit_size(size + 3);
+
+        prover.preprocess(&ck)?;
+        verifier.preprocess(&ck)?;
+
+        let pk = prover
+            .prover_key
+            .ok_or(anyhow!("Prover key is missing after preprocessing!"))?;
+
+        let vk = verifier
+            .verifier_key
+            .ok_or(anyhow!("Verifier key is missing after preprocessing!"))?;
+
+        self.set_prover_key(pk);
+        self.set_verifier_key(vk);
+
+        Ok(())
+    }
 
     /// Build PI vector for Proof verifications.
-    fn build_pi(&self, pub_inputs: &[PublicInput]) -> Result<Vec<BlsScalar>>;
+    fn build_pi(&self, pub_inputs: &[PublicInput]) -> Vec<BlsScalar> {
+        let mut pi = vec![BlsScalar::zero(); self.circuit_size()];
+
+        pub_inputs.iter().for_each(|p| match p {
+            PublicInput::BlsScalar(s, pos) => pi[*pos] = *s,
+            PublicInput::JubJubScalar(s, pos) => pi[*pos] = BlsScalar::from(*s),
+            PublicInput::AffinePoint(p, pos_x, pos_y) => {
+                pi[*pos_x] = p.get_x();
+                pi[*pos_y] = p.get_y();
```

Comment:
Please copy the same code that I did on the original implementation. X and Y have to be negated here. Otherways, this will make all of our circuits to fail.

Reply:
Done